go/vt/vtgateproxy/firstready_balancer.go

```diff
@@ -0,0 +1,80 @@
+package vtgateproxy
+
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// The firstready balancer implements the GRPC load balancer abstraction by
+// routing all queries to the first available target in the list returned from
+// discovery.
+//
+// Similar to the builtin "round_robin" balancer, the base functionality takes care
+// of establishing subconns to all targets in the list and keeping them in the "ready"
+// state, so all we have to do is pick the first available one from the set.
+//
+// This is in contrast to the `pick_first` balancer which only establishes the subconn
+// to a single target at a time and is therefore subject to undesirable behaviors if,
+// for example, the first host in the set is unreachable for some time, but not declared
+// down.
+// https://github.com/grpc/grpc-go/blob/master/pickfirst.go
+
+import (
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/grpclog"
+)
+
+// Name is the name of first_ready balancer.
+const Name = "first_ready"
+
+var logger = grpclog.Component("firstready")
+
+// newBuilder creates a new roundrobin balancer builder.
+func newBuilder() balancer.Builder {
+	return base.NewBalancerBuilder(Name, &frPickerBuilder{}, base.Config{HealthCheck: true})
+}
+
+func init() {
+	balancer.Register(newBuilder())
+}
+
+type frPickerBuilder struct{}
+
+func (*frPickerBuilder) Build(info base.PickerBuildInfo) balancer.Picker {
+	logger.Infof("firstreadyPicker: Build called with info: %v", info)
+	if len(info.ReadySCs) == 0 {
+		return base.NewErrPicker(balancer.ErrNoSubConnAvailable)
+	}
+
+	var subConn balancer.SubConn
+	for sc := range info.ReadySCs {
+		subConn = sc
```

Comment:
Note that we might want to actually pick _randomly_ from the set of ready subconns rather than just taking the first one in the map. 

I'm frankly not sure it matters in practice since the order should be basically nondeterministic anyway, and this is downstream from the already randomized shuffle from discovery, but in case it's possible that something like the order in which vtgates are started up somehow influences this, then maybe there's extra value in doing so?

Reply:
I think we want this to be deterministic in the input - we _assume_ that `Build` will be called only when there is a change in the ready list, but maybe there are circumstances where nothing has changed but it gets re-invoked. In those situations we don't want to jump around vtgates. (This is different from `round_robin` which explicitly does want to choose a different starting point each time, and has code to do that).

Reply:
Thinking further - we only want to pick a new gate if the currently picked one is no longer ready. So the Nth subconn changing state isn't something we should act on. We should pick the same subconn if it's still in the ready list, which is tricky because we don't know what we last picked. 

Without this, any vtgate subconn state change causes us to pick a new gate. We'd need some stable, but host-specific, way of deterministically choosing the same gate from a list (like sorting by hash of `<conn-id>-<hostname>`.

Reply:
If deterministic order matters, we can't rely on the ordering of this map, since map ordering is not guaranteed (in past versions of go it was intentionally shuffled). We would need to take the output of this map and sort it deterministically. If we don't care about order, then this is fine, but we'll have to be aware of the fact that the request may or may not change during every discovery.